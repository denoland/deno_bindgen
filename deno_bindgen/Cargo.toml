[package]
name = "deno_bindgen"
version = "0.8.1"
description = "This tool aims to simplify glue code generation for Deno FFI libraries written in Rust."
documentation = "https://docs.rs/deno_bindgen"
homepage = "https://github.com/denoland/deno_bindgen"
repository = "https://github.com/denoland/deno_bindgen"
keywords = ["deno", "ffi", "bindgen", "bindings", "macro"]
categories = ["development-tools::ffi", "development-tools"]
readme = "../README.md"
license = "MIT"
edition = "2021"

[lib]
path = "./lib.rs"

[dependencies]
deno_bindgen_macro = { path = "../deno_bindgen_macro", version = "0.8.1" }
serde = { version = "1", features = ["derive"] }
serde_json = "1"
