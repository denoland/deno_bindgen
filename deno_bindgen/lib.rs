// Copyright 2020-2021 the Deno authors. All rights reserved. MIT license.

//! ## deno_bindgen
//! This tool aims to simply types & glue code generation for FFI
//! libraries written in Rust.
//!
//! ### Usage
//! Add `serde` and `deno_bindgen` dependency to your crate.
//!
//! ```
//! use deno_bindgen::deno_bindgen;
//!
//! #[deno_bindgen]
//! pub struct Input {
//!   /// Doc comments are transformed into
//!   /// jsdocs.
//!   a: Vec<Vec<String>>,
//! }
//!
//! #[deno_bindgen(non_blocking)]
//! pub fn say_hello(message: &str) {
//!   println!("{}", message);
//! }
//! ```
//!
//! Generated bindings will look like this:
//! ```
//! // bindings/binding.ts
//!
//! // ... <init code here>
//!
//! type Input = {
//!   /**
//!    * Doc comments are transformed into
//!    * jsdocs.
//!    **/
//!   a: Array<Array<string>>;
//! };
//!
//! export async function say_hello(message: string) {
//!   // ... <glue code for symbol here>
//! }
//! ```
//! These bindings contain nessecary code to open the shared library,
//! define symbols and expose type definitions.
//! They can be simply imported into Deno code:
//! ```
//! import { say_hello } from "./bindings/bindings.ts";
//! await say_hello("Demn!")
//! ```
//!
pub use ::serde_json;
pub use deno_bindgen_macro::deno_bindgen;
