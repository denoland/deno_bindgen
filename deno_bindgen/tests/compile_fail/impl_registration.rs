use deno_bindgen::deno_bindgen;

// struct Foo is not "registered" in the inventory, so it `impl Foo`
// is not allowed.
struct Foo;

#[deno_bindgen]
impl Foo {
  #[constructor]
  fn new() -> Foo {
    Foo
  }
}

fn main() {}