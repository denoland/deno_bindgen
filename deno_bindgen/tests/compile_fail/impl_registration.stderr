error[E0277]: the trait bound `Foo: BindgenType` is not satisfied
 --> tests/compile_fail/impl_registration.rs:8:6
  |
8 | impl Foo {
  |      ^^^ the trait `BindgenType` is not implemented for `Foo`
  |
note: required by a bound in `_assert_impl`
 --> tests/compile_fail/impl_registration.rs:7:1
  |
7 | #[deno_bindgen]
  | ^^^^^^^^^^^^^^^ required by this bound in `_assert_impl`
  = note: this error originates in the attribute macro `deno_bindgen` (in Nightly builds, run with -Z macro-backtrace for more info)
